contracts/crypto-verify/src/ethereum.rs

```diff
@@ -0,0 +1,170 @@
+use cosmwasm_std::{Api, StdError, StdResult};
+use rlp::RlpStream;
+use sha3::{Digest, Keccak256};
+
+#[allow(clippy::too_many_arguments)]
+pub fn verify_transaction(
+    api: &dyn Api,
+    from: [u8; 20],
+    to: [u8; 20],
+    nonce: u64,
+    gas: u128,
+    gas_price: u128,
+    value: u128,
+    data: &[u8],
+    chain_id: u64,
+    r: &[u8],
+    s: &[u8],
+    v: u64,
+) -> StdResult<bool> {
+    let sign_bytes =
+        serialize_unsigned_transaction(to, nonce, gas, gas_price, value, data, chain_id);
+    let hash = Keccak256::digest(&sign_bytes);
+    let mut rs: Vec<u8> = Vec::with_capacity(64);
+    rs.resize(32 - r.len(), 0); // Left pad r to 32 bytes
+    rs.extend_from_slice(r);
+    rs.resize(32 + (32 - s.len()), 0); // Left pad s to 32 bytes
+    rs.extend_from_slice(s);
+
+    let recovery = get_recovery_param(v, chain_id)?;
+    let calculated_pubkey = api.secp256k1_recover_pubkey(&hash, &rs, recovery)?;
+    let calculated_address = ethereum_address_raw(&calculated_pubkey)?;
+    if from != calculated_address {
+        return Ok(false);
+    }
+    let valid = api.secp256k1_verify(&hash, &rs, &calculated_pubkey)?;
+    Ok(valid)
+}
+
+fn serialize_unsigned_transaction(
+    to: [u8; 20],
+    nonce: u64,
+    gas_limit: u128,
+    gas_price: u128,
+    value: u128,
+    data: &[u8],
+    chain_id: u64,
+) -> Vec<u8> {
+    let mut stream = RlpStream::new();
+    stream.begin_list(9);
+    stream.append(&nonce);
+    stream.append(&gas_price);
+    stream.append(&gas_limit);
+    stream.append(&to.as_ref());
+    stream.append(&value);
+    stream.append(&data);
+    stream.append(&chain_id);
+    stream.append(&Vec::<u8>::new()); // empty r
+    stream.append(&Vec::<u8>::new()); // empty s
```

Comment:
Done in fe070d113de8c5cd715f19d183e5dacdb899d3c4. I don't know any proper spec of that.

Reply:
I see, of course, the empty `r` and `s` values are because you cannot sign a transaction that includes something you still don't have (the signature itself).

Then you use the same encoding / helper to build the signed transaction. Makes sense.

Reply:
Yeah. It is strange that you need those empty fields in the RLP encoded list for the unsigned transaction. But you do. A 7 element list is invalid. 🤷‍♂️

Reply:
Ethereum has some quirks, I guess. This is one, and the ridiculous prefix for the messages, that's another.